File: repertoire/data-integration/providers/transform_plugin/type_cast.ts
// Transform Plugin Provider: type_cast
// Cast values between types (string, number, boolean, timestamp).
// See Architecture doc for transform plugin interface contract.

export const PROVIDER_ID = 'type_cast';
export const PLUGIN_TYPE = 'transform_plugin';

export interface TransformConfig {
  options?: Record<string, unknown>;
}

export interface TypeSpec {
  type: string;
  nullable?: boolean;
}

export class TypeCastTransformProvider {
  transform(value: unknown, config: TransformConfig): unknown {
    if (value === null || value === undefined) {
      return null;
    }

    const targetType = (config.options?.targetType as string) ?? 'string';

    switch (targetType) {
      case 'string':
        return String(value);

      case 'number': {
        if (typeof value === 'number') return value;
        const str = String(value).trim();
        if (str === '') return null;
        const floatVal = parseFloat(str);
        if (isNaN(floatVal)) {
          throw new Error(`Cannot cast "${str}" to number`);
        }
        // Return integer if the string looks like an integer
        if (Number.isInteger(floatVal) && !str.includes('.')) {
          return parseInt(str, 10);
        }
        return floatVal;
      }

      case 'boolean': {
        if (typeof value === 'boolean') return value;
        const boolStr = String(value).trim().toLowerCase();
        if (['true', '1', 'yes', 'on'].includes(boolStr)) return true;
        if (['false', '0', 'no', 'off'].includes(boolStr)) return false;
        throw new Error(`Cannot cast "${value}" to boolean`);
      }

      case 'timestamp': {
        if (typeof value === 'number') return value;
        const dateStr = String(value).trim();
        const parsed = Date.parse(dateStr);
        if (isNaN(parsed)) {
          throw new Error(`Cannot cast "${dateStr}" to timestamp`);
        }
        return parsed;
      }

      case 'integer': {
        if (typeof value === 'number') return Math.trunc(value);
        const intStr = String(value).trim();
        const intVal = parseInt(intStr, 10);
        if (isNaN(intVal)) {
          throw new Error(`Cannot cast "${intStr}" to integer`);
        }
        return intVal;
      }

      case 'float': {
        if (typeof value === 'number') return value;
        const fStr = String(value).trim();
        const fVal = parseFloat(fStr);
        if (isNaN(fVal)) {
          throw new Error(`Cannot cast "${fStr}" to float`);
        }
        return fVal;
      }

      default:
        throw new Error(`Unsupported target type: ${targetType}`);
    }
  }

  inputType(): TypeSpec {
    return { type: 'any', nullable: true };
  }

  outputType(): TypeSpec {
    return { type: 'any', nullable: true };
  }
}

export default TypeCastTransformProvider;

File: repertoire/data-integration/providers/transform_plugin/TypeCast.swift
// Transform Plugin Provider: type_cast
// Cast values between types (string, number, boolean, timestamp).
// See Architecture doc for transform plugin interface contract.

import Foundation

public struct TransformConfig {
    public var options: [String: Any]
    public init(options: [String: Any] = [:]) {
        self.options = options
    }
}

public struct TypeSpec {
    public let type: String
    public let nullable: Bool
    public init(type: String, nullable: Bool = false) {
        self.type = type
        self.nullable = nullable
    }
}

public enum TransformError: Error {
    case invalidCast(String)
    case unsupportedType(String)
}

public final class TypeCastTransformProvider {
    public static let providerId = "type_cast"
    public static let pluginType = "transform_plugin"

    public init() {}

    public func transform(value: Any, config: TransformConfig) throws -> Any {
        if value is NSNull {
            return NSNull()
        }

        let targetType = (config.options["targetType"] as? String) ?? "string"

        switch targetType {
        case "string":
            return String(describing: value)

        case "number", "float":
            if let num = value as? Double { return num }
            if let num = value as? Int { return Double(num) }
            let str = String(describing: value).trimmingCharacters(in: .whitespaces)
            guard !str.isEmpty, let parsed = Double(str) else {
                throw TransformError.invalidCast("Cannot cast \"\(str)\" to number")
            }
            return parsed

        case "integer":
            if let num = value as? Int { return num }
            if let num = value as? Double { return Int(num) }
            let str = String(describing: value).trimmingCharacters(in: .whitespaces)
            guard !str.isEmpty, let parsed = Int(str) else {
                throw TransformError.invalidCast("Cannot cast \"\(str)\" to integer")
            }
            return parsed

        case "boolean":
            if let b = value as? Bool { return b }
            let str = String(describing: value).trimmingCharacters(in: .whitespaces).lowercased()
            switch str {
            case "true", "1", "yes", "on": return true
            case "false", "0", "no", "off": return false
            default:
                throw TransformError.invalidCast("Cannot cast \"\(value)\" to boolean")
            }

        case "timestamp":
            if let num = value as? Double { return num }
            if let num = value as? Int { return Double(num) }
            let str = String(describing: value).trimmingCharacters(in: .whitespaces)
            let formatter = ISO8601DateFormatter()
            formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            if let date = formatter.date(from: str) {
                return date.timeIntervalSince1970 * 1000
            }
            // Try without fractional seconds
            formatter.formatOptions = [.withInternetDateTime]
            if let date = formatter.date(from: str) {
                return date.timeIntervalSince1970 * 1000
            }
            let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = "yyyy-MM-dd"
            if let date = dateFormatter.date(from: str) {
                return date.timeIntervalSince1970 * 1000
            }
            throw TransformError.invalidCast("Cannot cast \"\(str)\" to timestamp")

        default:
            throw TransformError.unsupportedType("Unsupported target type: \(targetType)")
        }
    }

    public func inputType() -> TypeSpec {
        return TypeSpec(type: "any", nullable: true)
    }

    public func outputType() -> TypeSpec {
        return TypeSpec(type: "any", nullable: true)
    }
}
